content/chapters/compute/lecture/slides.md

```diff
@@ -0,0 +1,1161 @@
+ 2582   914  3376  82956  1-04:15:56 Ss    \_ gpg-agent
+19812     1   176    988    09:03:34 Ss   init
+19813 19812   176    988    09:03:34 S     \_ init
+19814 19813  6420  11256    09:03:34 Ss        \_ bash
+21713 19814  3132  10696       00:00 R+            \_ ps
+20650     1   176    988    05:43:41 Ss   init
+20651 20650   176    988    05:43:41 S     \_ init
+20652 20651   604   2612    05:43:41 Ss+       \_ sh
+20653 20652   536   2612    05:43:41 S+            \_ sh
+20658 20653   604   2612    05:43:41 S+                \_ sh
+20661 20658 74072 931768    05:43:41 Sl+                   \_ node
+20681 20661 44316 614424    05:43:40 Sl+                       \_ node
+20722 20661 39804 828152    05:43:40 Sl+                       \_ node
+20733 20661 147004 845832   05:43:40 Sl+                       \_ node
+20796 20733 119336 959952   05:43:38 Sl+                           \_ node
+21149 20733 41080 581504    04:31:20 Sl+                           \_ node
+20672     1   176    988    05:43:40 Ss   init
+20673 20672   176    988    05:43:40 S     \_ init
+20674 20673 41556 585172    05:43:40 Ssl+      \_ node
+20701     1   176    988    05:43:40 Ss   init
+20702 20701   176    988    05:43:40 S     \_ init
+20703 20702 48060 590308    05:43:40 Ssl+      \_ node
+20915     1   176    988    05:24:46 Ss   init
+20916 20915   176    988    05:24:46 S     \_ init
+20917 20916  5268  10188    05:24:46 Ss        \_ bash
+20964 20917 19840 516296    05:24:32 S+            \_ python3
+20981     1   176    988    05:22:54 Ss   init
+20982 20981   176    988    05:22:54 S     \_ init
+20983 20982  5100  10056    05:22:54 Ss+       \_ bash
+```
+
+----
+
+### Process Attributes - from Kernel Space
+
+* PCB (Process Control Block):
+    * Data structure in the kernel - `task_struct`
+    * All information about the process (from `ps`, `/proc`):
+
+----
+
+### Process Attributes - from Kernel Space
+
+The most important fields in `task_struct`:
+```c [2 | 4 | 6 | 8 - 9 | 11 - 13 | 15 - 16 | 18 - 19 | 21 | 23 | 25 - 26 | 28 - 29]
+struct task_struct {
+        unsigned int                    __state;
+
+        void                           *stack;
+
+        unsigned int                    flags;
+
+        int                             on_cpu;
+        int                             prio;
+
+        /* Scheduler information */
+        struct sched_entity             se;
+        const struct sched_class        *sched_class;
+
+        /* The VAS: memory mappings */
+        struct mm_struct                *mm;
+
+        int                             exit_state;
+        int                             exit_code;
+
+        pid_t                           pid;
+
+        struct task_struct __rcu        *parent;
+
+        /* Child processes */
+        struct list_head                children;
+
+        /* Open file information */
+        struct files_struct             *files;
+};
+```
+
+---
+
+### `fork`
```

Comment:
I think this is too sudden. I would add a section called "Process Lifetime". I would state the main phases of a process and user actions: creation / birth, runtime (and live investigation by the user via ps, top etc.), termination. And the actions done by the parent process corresponding to this.

I can easily see a diagram for this.

You can then do a slide for each state with a short summary. Then the next section would be creation / birth / fork / spawn.

Reply:
Even more so, I think fork is too much here, it's more of an internals aspect. I would first go discussing more on how you start processes, using an executable, understanding the interface.

See the demos:
```
razvan@yggdrasil:~/school/so/curs.git$ git remote show origin
* remote origin
  Fetch URL: so@elf.cs.pub.ro:git-repos/curs.git
  Push  URL: so@elf.cs.pub.ro:git-repos/curs.git
  HEAD branch: master
  Remote branch:
    master tracked
  Local branch configured for 'git pull':
    master merges with remote master
  Local ref configured for 'git push':
    master pushes to master (up to date)
razvan@yggdrasil:~/school/so/curs.git$ ls
curs-01-demo  curs-04-demo  curs-07-demo      curs-09-demo      curs-11-demo  exec-process-demo      make-clean    publish-demos
curs-02-demo  curs-05-demo  curs-07-demo.zip  curs-09-demo.zip  curs-12-demo  exec-process-demo.zip  mem-sec.wiki  sync.wiki
curs-03-demo  curs-06-demo  curs-08-demo      curs-10-demo      curs-13-demo  exec-process.wiki      note-curs
razvan@yggdrasil:~/school/so/curs.git$ ls curs-04-demo/
nice  pipe  pipe-exec  utils  yield
razvan@yggdrasil:~/school/so/curs.git$ ls curs-03-demo/
ctxt-switch  executor  fork-file-pointer  orphan-zombie  utils
razvan@yggdrasil:~/school/so/curs.git$ ls curs-03-demo/executor/
executor       executor_ansi.c  executor.c  executor.py     executor_spawn.c  executor_system    Makefile  printer.d
executor_ansi  executor_ansi.o  executor.o  executor_spawn  executor_spawn.o  executor_system.c  printer   printer.o
```

Also check the recording of my process lecture for a narrative: https://ctipub.sharepoint.com/:f:/s/03-ACS-L-A3-S2-SO-CA-CB-CC-2021/Epei_BOAME9PkX5jLuCl1owBj3upEwzHxDHQU-014wRHqA?e=Bugjvv